datashader/pipeline.py

```diff
@@ -1,77 +1,58 @@
-default imperative interface provided by other files.
-"""
-
-import param
+from toolz import identity
 
 from . import transfer_functions as tf
 from . import reductions
 from . import core
-from . import glyphs
 
-class Pipeline(param.Parameterized):
-    """
-    Configurable datashading pipeline.  Allows each element of the
-    pipeline to be specified independently without code duplication,
-    e.g. to show the effect of varying that element while keeping the
-    rest of the pipeline constant.
 
-    Given a dataframe-like object df, the supported pipeline is roughly:
-
-    1. create canvas of the requested size (in data space) and resolution
-    2. aggregate df into pixel-shaped bins using the glyph and agg specifications
-    3. apply specified transfer_fns, if any, in order, on the set of bin values
-    4. apply specified color_fn to translate each resulting bin value into a color
-    5. return the result as an image
-    """
+class Pipeline(object):
+    """A datashading pipeline callback.
 
-    df = param.Parameter(doc="""
-        Object supporting columnar-style data access, such as a Pandas
-        dataframe.""")
+    Given a declarative specification, creates a callable with the following
+    signature:
 
-    glyph = param.ClassSelector(glyphs.Glyph,default=glyphs.Point("x","y"), doc="""
-        Marker shape for each point, specified using fields for the data object.""")
+    ``callback(x_range, y_range, width, height)``
 
-    agg = param.ClassSelector(reductions.Reduction, default=reductions.count(), 
-        doc="""Function for incrementally reducing a bin's values into a scalar.""")
+    where ``x_range`` and ``y_range`` form the bounding box on the viewport,
+    and ``width`` and ``height`` specify the output image dimensions.
 
-    transfer_fns = param.HookList(default=[], doc="""
-        Optional function(s) to apply to the aggregated bin values, before
-        they each get converted into a color.""")
-
-    color_fn = param.Callable(default=tf.interpolate, doc="""
-        Function to convert a scalar aggregated bin value into a color.""")
-
-
-    def __call__(self, x_range, y_range, w, h, **params):
-        """
-        Accepts a viewport in data space specified as 
-        x_range (xmin,xmax), y_range (ymin,ymax), and a rendering
-        resolution w x h.
-
-        Returns an image of the specified height and width, rendered
-        over the specified range in data space, using the current
-        parameter values.
+    Parameters
+    ----------
+    df : pandas.DataFrame, dask.DataFrame
+    glyph : Glyph
+        The glyph to bin by.
+    agg : Reduction, optional
+        The reduction to compute per-pixel. Default is ``count()``.
+    transform_fn : callable, optional
+        A callable that takes the computed aggregate as an argument, and
+        returns another aggregate. This can be used to do preprocessing before
+        passing to the ``color_fn`` function.
```

Comment:
Having it be a list of 1 argument functions was fine, but a bit limiting. I'd argue that taking in just a single function (which could be composed externally) made more sense.